[Angelyxx] iP

Angelyxx opened the pull request:
### DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ TO USE

All you need to do is,

1. download it from [here](https://github.com/Angelyxx/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Angelyxx committed:
Level-2

Angelyxx committed:
Level-3

Angelyxx committed:
Level-4

Angelyxx committed:
Level-5

Angelyxx committed:
Level-6

Angelyxx committed:
A-TextUiTesting

Angelyxx committed:
Add classes to packages

Angelyxx committed:
complete level-7

Angelyxx committed:
allow duke to read date/dateTime object for deadline

Angelyxx committed:
allow datetime objects for event class

Angelyxx committed:
A-MoreOOP

Angelyxx committed:
Merge branch 'master' into add-gradle-support

Angelyxx committed:
remove duplicate files

Angelyxx committed:
Add JUnit Tests for ParserTest, StorageTest, TaskListTest

Angelyxx committed:
Added JavaDocs to all classes

Angelyxx committed:
Edit code to follow coding conventions

Angelyxx committed:
Complete find command

Angelyxx committed:
Add DukeException class

Angelyxx committed:
Merge remote-tracking branch 'origin/A-JavaDoc'

Angelyxx committed:
Merge remote-tracking branch 'origin/A-CodingStandard'

# Conflicts:
#	src/main/java/duke/main/Parser.java
#	src/main/java/duke/main/Storage.java
#	src/main/java/duke/main/TaskList.java
#	src/main/java/duke/main/Ui.java
#	src/main/java/duke/task/Task.java

Angelyxx committed:
Merge remote-tracking branch 'origin/Level-9'

# Conflicts:
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/main/Parser.java
#	src/main/java/duke/main/Ui.java

Angelyxx committed:
Fix TaskList typo error

jason-raiin commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class EventCommand extends Command {
+    private String description;
+    private String startDateTimeStr;
+    private String endDateTimeStr;
```
Minor nitpick: Not necessary to include type in variable names in a strongly typed language such as Java

jason-raiin commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,44 @@
+        if (index >= 0 && index < tasks.size()) {
+            Task markTask = tasks.get(index);
+            tasks.markDone(index);
+            ui.printMessage("Nice! I've marked this task as done:\n\t",markTask);
```
Missing space in method params. Recommend using a linter to automatically correct these minor issues

jason-raiin commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,44 @@
+        if (index >= 0 && index < tasks.size()) {
+            Task unmarkTask = tasks.get(index);
+            tasks.markNotDone(index);
+            ui.printMessage("OK, I've marked this task as not done yet:\n\t",unmarkTask);
```
missing space

jason-raiin commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,66 @@
+                ui.showError(e.getMessage());
+            } finally {
+                ui.showLine();
+            }
```
nice usage of try-catch-finally

TohLiYuan replied:
Good exception handling

jason-raiin commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,167 @@
+            return new FindCommand(keyword);
+        } else {
+            throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
```
a switch statement could help make this easier to read

TohLiYuan replied:
A switch case would make this more readable
Would it be better to extract the phrases out and avoid hardcoding into the if-else/switch-case statements?

jason-raiin commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,167 @@
+     * @throws DukeException If the input format is invalid.
+     */
+    private static String[] validateDeadline(String input) throws DukeException {
+        String[] splitInput = input.replace("deadline", "").split(" /by ");
```
`input.replace(...)` could be a potential bug, as it removes all instances of the string in the user input

TohLiYuan replied:
Replacing the string might cause unexpected issues, eg. when "deadline" is part of the name of the Deadline Task. This can be seen in the validateEvent as well

jason-raiin commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,118 @@
+ */
+
+public class Ui {
+    private Scanner sc;
```
imo abbreviating `scanner` is not terrible but also not ideal

jason-raiin commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,118 @@
+        formatters.add(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+        formatters.add(DateTimeFormatter.ofPattern("dd/MM/yyyy"));
+        formatters.add(DateTimeFormatter.ofPattern("MM/dd/yyyy"));
+        formatters.add(DateTimeFormatter.ofPattern("d/M/yyyy"));
```
`formatters` is a constant and should be in capitalized `SNAKE_CASE`. 

also probably more efficient to define this somewhere else so that you do not have to repeat this in both `Deadline` and `Event`, perhaps as a static variable in `Task`

TohLiYuan replied:
Good use of formatters list to specify different formattings of date time.
Perhaps consider making this an additional Date Parser class to be used globally?

jason-raiin commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+
+    public String toFileString() {
+        String type = "T"; // Default type for base Task
+        return type + " | " + (isDone ? "1" : "0") + " | " + description;
```
Use `String.format()` to increase readability of multiple string concatenations

jason-raiin reviewed (commented):
Overall LGTM. Some minor optimizations can be performed but are not strictly necessary.

TohLiYuan reviewed (commented):
Good readability overall, keep up the good work!

Angelyxx committed:
add GUI to Duke

Angelyxx committed:
add assertions

Angelyxx committed:
Merge pull request #1 from Angelyxx/A-Assertions

add assertions

Angelyxx committed:
Improve code quality by following java coding conventions

Angelyxx committed:
Debug validate methods for todo, deadline and event commands

Previous implementation did not allow for users to include the words "todo", "deadline" and "event" in their description as method used to replace those words for their respective commands.

Adjusting the validation to allow for those words to be included in the description results in better code quality and flexibility.

Angelyxx committed:
Merge pull request #2 from Angelyxx/A-CodeQuality

A-code quality

Angelyxx committed:
Allow for more natural formats of date inputs

This will allow for users to be more flexible in their datetime inputs for their deadline and todo tasks.

Let's put all date and time parsing methods and dateTime formatters into the dateParser class.

Having a dateParser utility class allows us to use its methods across different classes, for example date parsing can be done in both Event and Deadline classes.

Angelyxx committed:
Merge pull request #3 from Angelyxx/BCD-Extension

Allow for more natural formats of date inputs

Angelyxx committed:
Clean up code

Angelyxx committed:
debugged validation of find command

Angelyxx committed:
Update README.md

Angelyxx committed:
Update README.md

Angelyxx committed:
Update README.md

Angelyxx committed:
Update README.md

Angelyxx committed:
Update README.md

Angelyxx committed:
Update README.md

Angelyxx committed:
Edit parser commands

Angelyxx committed:
debug gradle file to pass smoke test

Angelyxx committed:
Update README.md
